### PR TITLE
Implement predicate functions: all(), any(), none(), single()

### LINKED ISSUE #552: Implement predicate function: all
**Is your feature request related to a problem? Please describe.**
Descriptions of this function is exactly similar to the one described here: https://neo4j.com/docs/cypher-manual/current/functions/predicate.

**Describe the solution you'd like**
N/A

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### LINKED ISSUE #553: Implement predicate function: any
**Is your feature request related to a problem? Please describe.**
Descriptions of this function is exactly similar to the one described here: https://neo4j.com/docs/cypher-manual/current/functions/predicate.

**Describe the solution you'd like**
N/A

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### LINKED ISSUE #555: Implement predicate function: none
**Is your feature request related to a problem? Please describe.**
Descriptions of this function is exactly similar to the one described here: https://neo4j.com/docs/cypher-manual/current/functions/predicate.

**Describe the solution you'd like**
N/A

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### LINKED ISSUE #556: Implement predicate function: single
**Is your feature request related to a problem? Please describe.**
Descriptions of this function is exactly similar to the one described here: https://neo4j.com/docs/cypher-manual/current/functions/predicate.

**Describe the solution you'd like**
N/A

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by gregfelice**
## Summary

Implements the four openCypher predicate functions (issues #552, #553, #555, #556):

- `all(x IN list WHERE predicate)` — true if all elements match
- `any(x IN list WHERE predicate)` — true if at least one matches
- `none(x IN list WHERE predicate)` — true if no elements match
- `single(x IN list WHERE predicate)` — true if exactly one matches

These are among the most requested Cypher features for AGE and are critical for users migrating from Neo4j and Kuzu (recently archived).

## Implementation

**Approach:** Builds on the existing list comprehension infrastructure (`unnest`-based subqueries with child parsestates for variable scoping).

**SQL transformation strategy:** Each predicate function is lowered to an aggregate subquery over `unnest(list)` that preserves Cypher's three-valued (TRUE/FALSE/NULL) semantics:

- `all()`, `any()`, `none()` → `SELECT CASE WHEN bool_or(<first-branch>) THEN <result1> WHEN bool_or(pred IS NULL) THEN NULL ELSE <result2> END FROM unnest(list) AS x` — two `bool_or()` aggregates compute (1) whether any element satisfies the decisive condition and (2) whether any element's predicate is NULL, combined via `CASE` to return the correct three-valued result. On an empty list both aggregates return NULL, so the `ELSE` branch yields the vacuous-truth defaults (`all()`/`none()` → true, `any()` → false).
- `single()` → `SELECT count(*) FROM unnest(list) AS x WHERE pred IS TRUE) = 1` — exact count of truthy matches. (LIMIT 2 short-circuit optimization deferred; current form evaluates the full list.)
- All four are wrapped in a `CASE WHEN <list-expr> IS NULL THEN NULL ELSE <subquery> END` guard so a NULL list expression propagates NULL rather than collapsing to the empty-list defaults.

This aggregate-based shape (rather than `EXISTS`/`NOT EXISTS`) was chosen specifically to preserve correct NULL handling: `WHERE` in a SQL subquery drops rows where the predicate is NULL, which would incorrectly collapse `any(x IN [NULL] WHERE x > 0)` from NULL to false under an `EXISTS` shape. The aggregate form keeps NULL-producing rows visible and folds them into a proper three-valued result.

**Files changed (12):**

| File | Change |
|------|--------|
| `cypher_nodes.h` | New `cypher_predicate_function` node type with `CPFK_ALL/ANY/NONE/SINGLE` enum |
| `ag_nodes.h` / `ag_nodes.c` | Node registration |
| `cypher_outfuncs.h` / `cypher_outfuncs.c` | Serialization |
| `cypher_kwlist.h` | Three new keywords: `ANY_P`, `NONE`, `SINGLE` |
| `cypher_gram.y` | Grammar rules in `expr_func_subexpr`, `build_predicate_function_node()` helper with NULL-list guard, `extract_iter_variable_name()` shared helper (rejects qualified ColumnRefs), keywords added to `safe_keywords` |
| `cypher_clause.c` | `transform_cypher_predicate_function()` — builds query tree with `bool_or()` aggregates + `CASE` for three-valued semantics; `make_bool_or_agg()` helper |
| `cypher_analyze.c` | Expression walker for new node type |
| `Makefile` | Register new regression test |
| `regress/sql/predicate_functions.sql` | New test file |
| `regress/expected/predicate_functions.out` | Expected output |

**Backward compatibility:**
- `ANY_P`, `NONE`, `SINGLE` added to `safe_keywords` so they work as property keys and label names (e.g., `{any: 1, none: 2, single: 3}`)
- `ALL` was already a reserved keyword with `safe_keywords` entry
- No grammar conflicts (verified: zero new shift/reduce or reduce/reduce warnings)

## Regression Tests

Test queries covering:
- Basic true/false for each function
- Empty list edge cases (vacuous truth for `all()`/`none()`, false for `any()`/`single()`)
- NULL list input (all four return NULL via the guard)
- NULL predicate results within non-empty lists (three-valued semantics)
- Graph data integration (`MATCH (u) WHERE all(x IN u.vals WHERE ...)`)
- Boolean combinations (`any(...) AND all(...)`)
- Nested predicates (`any(x IN ... WHERE all(y IN ... WHERE ...))`)
- Keyword backward compatibility (`{any: 1, none: 2, single: 3}`)
- Deterministic ordering (all graph queries use `ORDER BY`)

All regression tests pass.

### CODE REVIEW

**Comment by jrgemignani** [comment]
@gregfelice Please see the above comments by Copilot

**Comment by gregfelice** [comment]
Addressed all 4 Copilot suggestions:

1. **NULL semantics** — Rewrote from EXISTS_SUBLINK to EXPR_SUBLINK with aggregate-based CASE expressions (`bool_or(pred IS TRUE/FALSE)` + `bool_or(pred IS NULL)`) to preserve three-valued Cypher NULL semantics. `any(x IN [NULL] WHERE x > 0)` now correctly returns NULL instead of false.

2. **Unqualified iterator check** — Added `list_length(cref->fields) != 1` validation in `extract_iter_variable_name()`. Qualified names like `x.y` now error with "qualified name not allowed as iterator variable".

3. **NULL regression tests** — Added 13 new test cases covering: NULL list input for all four functions, null elements in lists, literal null predicates, and mixed null/non-null elements.

4. **single() performance** — Applied `IS TRUE` filtering so NULL predicates aren't counted. The LIMIT 2 optimization breaks correlated variable references in graph property contexts (e.g., `MATCH (u) WHERE single(x IN u.vals WHERE ...)`), documented with TODO for future optimization pass.

**Bonus**: Added copy/read support for `cypher_predicate_function` ExtensibleNode to prevent "unexpected copyObject()" crashes when PostgreSQL's query rewriter copies expression trees.

All 32 regression tests pass (`predicate_functions`: ok).

**Comment by jrgemignani** [comment]
@gregfelice We'll see what Copilot thinks ;) Btw, in the future, can you put your comments in the reply to Copilot, please.

File: src/backend/parser/cypher_gram.y [hunk 1]
```diff
@@ -3317,6 +3382,79 @@ static Node *build_list_comprehension_node(Node *var, Node *expr,
...
+        return (Node *) node_to_agtype(eq_expr, "boolean", location);
+    }
+    else
+    {
+        /*
+         * all()/any()/none(): the subquery returns a boolean directly
+         * from the CASE+bool_or() aggregate expression.
+         */
+        return (Node *) node_to_agtype((Node *) sub, "boolean", location);
+    }
```
[the remark this answers is not in the document]
Comment (gregfelice) [reply src/backend/parser/cypher_gram.y:3488]: Addressed in cc19476b. Added a NULL-list guard in the grammar layer:

```c
CASE WHEN expr IS NULL THEN NULL ELSE <predicate-result> END
```

The guard wraps the entire predicate function result using a shared pointer to the list expression (safe because AGE's `transform_cypher_expr_recurse` creates new nodes without modifying the parse tree in-place). This fixes `single(x IN null WHERE ...)` returning false instead of NULL, and makes the NULL behavior explicit rather than relying on aggregate-over-empty-set semantics.

Updated regression tests to expect NULL for `single(x IN null WHERE ...)`.

File: src/backend/parser/cypher_clause.c [hunk 2]
```diff
@@ -1602,6 +1611,332 @@ static Query *transform_cypher_list_comprehension(cypher_parsestate *cpstate,
...
+ *   FROM unnest(list) AS x
+ *
+ * For single():
+ *   SELECT count(*) FROM (
+ *     SELECT 1 FROM unnest(list) AS x WHERE pred IS TRUE LIMIT 2
+ *   ) sub
+ *
+ * All four use EXPR_SUBLINK so the subquery returns a scalar value.
+ * The LIMIT 2 in single() enables early termination: once two matches
+ * are found, evaluation stops and count(*) = 2 != 1.
```
[the remark this answers is not in the document]
Comment (gregfelice) [reply src/backend/parser/cypher_clause.c]: Addressed in cc19476b. Updated the block comment to describe the actual implementation (plain count(*) without LIMIT 2). The LIMIT 2 optimization is noted as deferred in the single() branch comment.

File: src/backend/parser/cypher_clause.c [hunk 3]
```diff
@@ -25,6 +25,7 @@
 #include "postgres.h"
 
 #include "access/heapam.h"
+#include "catalog/pg_aggregate.h"
```
[the remark this answers is not in the document]
Comment (gregfelice) [reply src/backend/parser/cypher_clause.c:28]: Investigated — this include is NOT unused. `AGGKIND_NORMAL` (used in `make_bool_or_agg()` at line 1663) is defined in `catalog/pg_aggregate.h`. Build fails without it. Keeping as-is.

**Comment by jrgemignani** [comment]
@gregfelice Please see Copilot above

File: src/backend/parser/cypher_gram.y [hunk 4]
```diff
@@ -3268,6 +3295,44 @@ static cypher_relationship *build_VLE_relation(List *left_arg,
...
+                (errcode(ERRCODE_SYNTAX_ERROR),
+                 errmsg("invalid iterator variable name")));
+    }
+
+    return val->sval;
+}
+
 /* helper function to build a list_comprehension grammar node */
 static Node *build_list_comprehension_node(Node *var, Node *expr,
                                            Node *where, Node *mapping_expr,
```
[the remark this answers is not in the document]
Comment (gregfelice) [reply src/backend/parser/cypher_gram.y:3338]: Addressed in commit 507a2e53. `build_list_comprehension_node()` now calls `extract_iter_variable_name()` directly (cypher_gram.y:3352), so iterator-variable validation — including rejection of qualified ColumnRefs like `x.y IN list` — is shared between list comprehensions and predicate functions. The `list_comprehension` regression expected output was updated to the normalized lowercase error message.

File: src/backend/parser/cypher_clause.c [hunk 5]
```diff
@@ -1602,6 +1611,331 @@ static Query *transform_cypher_list_comprehension(cypher_parsestate *cpstate,
...
+ * NULL semantics.  The grammar layer wraps the SubLink with a
+ * CASE WHEN list IS NULL THEN NULL ELSE (subquery) END guard so all
+ * four functions return NULL when the input list is NULL.
+ *
+ * For all()/any()/none():
+ *   SELECT CASE WHEN bool_or(pred IS TRUE/FALSE) THEN ...
+ *               WHEN bool_or(pred IS NULL)       THEN NULL
+ *               ELSE ... END
+ *   FROM unnest(list) AS x
+ *
```
[the remark this answers is not in the document]
Comment (gregfelice) [reply src/backend/parser/cypher_clause.c:1789]: The PR description was updated to match the actual implementation — it now explicitly describes the `bool_or()` aggregate + `CASE` shape and calls out that this was chosen *rather than* EXISTS/NOT EXISTS, specifically to preserve correct three-valued NULL handling (WHERE in a SQL subquery drops NULL-producing rows, which would collapse `any(x IN [NULL] WHERE x > 0)` from NULL to false under an EXISTS shape). No short-circuiting is claimed. See the "SQL transformation strategy" section of the PR description.

**Comment by gregfelice** [comment]
@jrgemignani — all Copilot items addressed and threads resolved. Round 3 commit (507a2e53) refactors `build_list_comprehension_node()` to reuse `extract_iter_variable_name()`, so iterator-variable validation is consistent between list comprehensions and predicate functions. The PR description already describes the `bool_or()` aggregate approach correctly. Ready for re-review when you have a moment. Thanks!

**Review by jrgemignani — APPROVED** [review]

**Comment by jrgemignani** [comment]
@gregfelice I'm going to need you to rebase this due to an issue we had with the new makefile.

**Comment by gregfelice** [comment]
@jrgemignani — rebased onto current `master` (now includes #2397 makefile fix) and force-pushed (`507a2e53` → `143787fe`). Clean rebase, no conflicts. Local verification on PG18:

- `make clean && make -j8` — clean build, no warnings
- `pg_regress --load-extension=age predicate_functions list_comprehension` — both pass

Ready for merge whenever you are. Thanks!

**Review by MuhammadTahaNaveed — APPROVED** [review]
